### PR TITLE
Fix image orientation when generating thumbnail

### LINKED ISSUE #2384: Image thumbnailing doesn't preserve orientation
Hi,

When I upload image [1], which has the Exif orientation value of 3 (rotate 180°), the thumbnailed image [2] appears upside down :)

[1] https://omino.foldplop.com/~ahihi/IMG_1190.JPG
[2] https://omino.foldplop.com/~ahihi/MeJGeqvdznHgjTSVIQZYPkbf.jpeg

### PR DESCRIPTION
**Pull request by prodrigestivill**
Add EXIF rotation before generating thumbnails.

Fixes #2384, #1261, vector-im/riot-web#1999, and vector-im/riot-web#2059

For now it needs `pillow>=4.3.0`.
It can be moved to use `Image.getexif()` instead of `Image._getexif()` but it will require `pillow>=6.0.0`.
This is explained in [Pillow 6.0.0 Documentation](https://pillow.readthedocs.io/en/stable/releasenotes/6.0.0.html#added-exif-class).

Not sure if the preference is to use the oldest minimum libraries or no issues at all with using latest ones.

### Pull Request Checklist

* [X] Pull request is based on the develop branch
* [X] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [X] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

**Comment by prodrigestivill** [comment]
Afects matrix-org/matrix-doc#483

File: synapse/rest/media/v1/thumbnailer.py [hunk 1]
```diff
@@ -30,6 +30,26 @@ class Thumbnailer(object):
...
+                6: Image.ROTATE_270,
+                7: Image.TRANSVERSE,
+                8: Image.ROTATE_90
+            }.get(image_orientation)
+            if transpose_method is not None:
+                self.image = self.image.transpose(transpose_method)
+                # We don't need EXIF any more
+                self.image.info["exif"] = None
+        except Exception:
+            # A lot of parsing errors can happen when parsing EXIF
```
Comment (anoadragon453) [inline synapse/rest/media/v1/thumbnailer.py]: Probably worth adding a logline here just to help people in debugging purposes. Something like:

```
logger.debug("Error parsing image EXIF information: %s", e)
```

And changing the `except` line to:

```
except Exception as e:
```

to give you the exception.

**Comment by prodrigestivill** [comment]
> Looks good otherwise, but we'll need to see if this is the approach we want to take.

The main problem with current Thumbnailer is that creates an scaled or cropped thumbnail and throws away the EXIF orientation information.
So either rotate before the scale/crop or copy the EXIF orientation information when generating the thumbnail.

So for me this problem is independent of what is decided in matrix-org/matrix-doc#483 but it still affects to solve the whole problem.

But it can still be decied to copy the EXIF orientation in the thumbnail instead, but I think this would be an strange solution.

File: synapse/rest/media/v1/thumbnailer.py [hunk 2]
```diff
@@ -30,6 +30,26 @@ class Thumbnailer(object):
 
     def __init__(self, input_path):
         self.image = Image.open(input_path)
+        try:
+            # We don't use ImageOps.exif_transpose since it crashes with big EXIF
+            image_exif = self.image._getexif()
+            image_orientation = image_exif.get(0x0112)
```
Comment (richvdh) [inline synapse/rest/media/v1/thumbnailer.py]: could we use a symbolic constant here? https://pillow.readthedocs.io/en/3.0.x/reference/ExifTags.html

File: synapse/rest/media/v1/thumbnailer.py [hunk 3]
```diff
@@ -30,6 +30,26 @@ class Thumbnailer(object):
...
+                4: Image.FLIP_TOP_BOTTOM,
+                5: Image.TRANSPOSE,
+                6: Image.ROTATE_270,
+                7: Image.TRANSVERSE,
+                8: Image.ROTATE_90
+            }.get(image_orientation)
+            if transpose_method is not None:
+                self.image = self.image.transpose(transpose_method)
+                # We don't need EXIF any more
+                self.image.info["exif"] = None
```
Comment (richvdh) [inline synapse/rest/media/v1/thumbnailer.py]: this seems a bit brutal?

File: synapse/rest/media/v1/thumbnailer.py (hunk 2 above)
Comment (prodrigestivill) [reply synapse/rest/media/v1/thumbnailer.py]: The constants in PIL.ExifTags.TAGS are indexed reversed.
Looks like the dictionary is meant to display the values to the user instead of looking for them.

The value we want (`0x0112`) is the key if the dictionary, so this means we need to lookup by value (sub-optimal).

The value `0x0112` comes from EXIF standard, so its not PIL.ExifTags specific value that they can change.

In case we want to loopup by value, shoud we lookup for this key as a static code above instead of doing this each time?
Or maybe adding a comment in the same line would be enought?

Another option would be update to PIL 6 and try to fix ImageOps.exif_transpose for images with big EXIF.
The issue with the ImageOps.exif_transpose happends when they try to recreate the EXIF with no orientation tag after rotating, since we don't need EXIF at all this issue should not affect this project.

File: synapse/rest/media/v1/thumbnailer.py (hunk 3 above)
Comment (prodrigestivill) [reply synapse/rest/media/v1/thumbnailer.py]: In class Thumbnailer the main goal is to generate a Thumbnail.
Thumbnails normally don't have any EXIF information since the main goal of them are to weight less.

The current released version of this class is generating thumbnails without rotating and is droping all EXIF information (including the orientation tag) and this is what is causing the issue.

So since all EXIF information is currently droped when exporting the scaled image, we don't really need this information at all either for current code nor for this pull request code.

This line is only to avoid any future issue with the wrong orientation EXIF tag, because in this case the image is already rotated and the EXIF orientation tag would tell that needs to rotate.

The nice solution would be trying to remove only EXIF orientation tag.
This is what ImageOps.exif_transpose is trying to do but crashes with images with big EXIF (generated by Apple iPhone for example).

To summarize:
- Curent Thumbnailer class is dropping EXIF on export image.
- EXIF information is not needed once the image has been rotated.
- EXIF information can be droped for all images as soon as is loaded and rotated to free memory sooner, but this would make code more strange.

File: synapse/rest/media/v1/thumbnailer.py (hunk 2 above)
Comment (richvdh) [reply synapse/rest/media/v1/thumbnailer.py]: Yeah, neither of those are good solutions. Fair enough, let's stick with the hardcoded number.

File: synapse/rest/media/v1/thumbnailer.py [hunk 4]
```diff
@@ -30,6 +30,26 @@ class Thumbnailer(object):
...
+        except Exception as e:
+            # A lot of parsing errors can happen when parsing EXIF
+            logger.debug("Error parsing image EXIF information: %s", e)
```
Comment (richvdh) [inline synapse/rest/media/v1/thumbnailer.py]: 1. I suggest making this at least an `info`, since most people have `debug` turned off.
2. We won't just reach this point for errors parsing the EXIF, but also when doing the transpose. We need to either make the `try` clause cover less code, or fix the log message (and comment) to better reflect the operations that might have caused the exception.

File: synapse/rest/media/v1/thumbnailer.py (hunk 3 above)
Comment (richvdh) [reply synapse/rest/media/v1/thumbnailer.py]: You're right.

When I first looked at this code, I assumed that we were stripping the EXIF information from the *original* image. (There may be valid reasons to do that, but it's a thing that we'd need to think about.) However, I now see that the original image is not affected. So this is fine :+1:

File: synapse/rest/media/v1/thumbnailer.py [hunk 5]
```diff
@@ -30,6 +30,26 @@ class Thumbnailer(object):
...
+                2: Image.FLIP_LEFT_RIGHT,
+                3: Image.ROTATE_180,
+                4: Image.FLIP_TOP_BOTTOM,
+                5: Image.TRANSPOSE,
+                6: Image.ROTATE_270,
+                7: Image.TRANSVERSE,
+                8: Image.ROTATE_90
+            }.get(image_orientation)
+            if transpose_method is not None:
+                self.image = self.image.transpose(transpose_method)
```
[the remark this answers is not in the document]
Comment (prodrigestivill) [reply synapse/rest/media/v1/thumbnailer.py]: I moved the code to allow it to be called outside Thumbnailer class.
Making it thread safe, simpler and allows it to be executed in the thread reactor.

**Comment by richvdh** [comment]
oh urgh the isort failure is something that is fixed in `develop`. Could you merge in latest develop? sorry, but the rest of the CI won't run until it's fixed.

**Comment by johansmitsnl** [comment]
Should this be working in the latest docker image? For me it is not working.

**Comment by prodrigestivill** [comment]
It should be working on latest official Docker image.

Please ensure you are looking at the thumbnail is generated by the sever and not the thumbnail generated by the client or the thumbnail of the thumbnail generated by the client.

Normally Matrix clients generate their own thumbnails and upload them without any EXIF information.
